sdk/core/azure-core/src/http/websockets/websockets_impl.cpp

```diff
@@ -0,0 +1,884 @@
+      }
+      return ss.str();
+    }
+  } // namespace
+
+  WebSocketImplementation::WebSocketImplementation(
+      Azure::Core::Url const& remoteUrl,
+      WebSocketOptions const& options)
+      : m_remoteUrl(remoteUrl), m_options(options), m_pingThread(this, m_options.PingInterval)
+  {
+  }
+
+  void WebSocketImplementation::Open(Azure::Core::Context const& context)
+  {
+    if (m_state != SocketState::Invalid && m_state != SocketState::Closed)
+    {
+      throw std::runtime_error("Socket is not closed.");
+    }
+    m_state = SocketState::Opening;
+
+#if defined(BUILD_TRANSPORT_WINHTTP_ADAPTER)
+    WinHttpTransportOptions transportOptions;
+    auto winHttpTransport
+        = std::make_shared<Azure::Core::Http::WebSockets::WinHttpWebSocketTransport>(
+            transportOptions);
+    m_transport = std::static_pointer_cast<WebSocketTransport>(winHttpTransport);
+    m_options.Transport.Transport = std::static_pointer_cast<HttpTransport>(winHttpTransport);
+#elif defined(BUILD_CURL_HTTP_TRANSPORT_ADAPTER)
+    CurlWebSocketTransportOptions transportOptions;
+    transportOptions.HttpKeepAlive = false;
+    auto curlWebSockets
+        = std::make_shared<Azure::Core::Http::WebSockets::CurlWebSocketTransport>(transportOptions);
+
+    m_transport = std::static_pointer_cast<WebSocketTransport>(curlWebSockets);
+    m_options.Transport.Transport = std::static_pointer_cast<HttpTransport>(curlWebSockets);
+#endif
+
+    std::vector<std::unique_ptr<Azure::Core::Http::Policies::HttpPolicy>> perCallPolicies{};
+    std::vector<std::unique_ptr<Azure::Core::Http::Policies::HttpPolicy>> perRetryPolicies{};
+    // If the caller has told us a service name, add the telemetry policy to the pipeline to add
+    // a user agent header to the request.
+    if (!m_options.ServiceName.empty())
+    {
+      perCallPolicies.push_back(
+          std::make_unique<Azure::Core::Http::Policies::_internal::TelemetryPolicy>(
+              m_options.ServiceName, m_options.ServiceVersion, m_options.Telemetry));
+    }
+    Azure::Core::Http::_internal::HttpPipeline openPipeline(
+        m_options, std::move(perRetryPolicies), std::move(perCallPolicies));
+
+    Azure::Core::Http::Request openSocketRequest(
+        Azure::Core::Http::HttpMethod::Get, m_remoteUrl, false);
+
+    // Generate the random request key. Only used when the transport doesn't support websockets
+    // natively.
+    auto randomKey = GenerateRandomKey();
+    auto encodedKey = Azure::Core::Convert::Base64Encode(randomKey);
+    if (!m_transport->HasBuiltInWebSocketSupport())
+    {
+      // If the transport doesn't support WebSockets natively, set the standardized WebSocket
+      // upgrade headers.
+      openSocketRequest.SetHeader("Upgrade", "websocket");
+      openSocketRequest.SetHeader("Connection", "upgrade");
+      openSocketRequest.SetHeader("Sec-WebSocket-Version", "13");
+      openSocketRequest.SetHeader("Sec-WebSocket-Key", encodedKey);
+    }
```

Comment:
```suggestion
    if (!m_transport->HasBuiltInWebSocketSupport())
    {
      // If the transport doesn't support WebSockets natively, set the standardized WebSocket
      // upgrade headers.
      auto randomKey = GenerateRandomKey();
      auto encodedKey = Azure::Core::Convert::Base64Encode(randomKey);

      openSocketRequest.SetHeader("Upgrade", "websocket");
      openSocketRequest.SetHeader("Connection", "upgrade");
      openSocketRequest.SetHeader("Sec-WebSocket-Version", "13");
      openSocketRequest.SetHeader("Sec-WebSocket-Key", encodedKey);
    }
```

Reply:
Unfortunately, RandomKey/encodedkey needs to be at the larger scope because they're referenced below the call to Send as a part of validating the response headers :(.